src/librustc_mir/interpret/eval_context.rs

```diff
@@ -1314,47 +1316,117 @@ impl<'a, 'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> EvalContext<'a, 'mir, 'tcx, M
         }
     }
 
+    fn validate_scalar(
+        &self,
+        value: Scalar,
+        size: Size,
+        scalar: &layout::Scalar,
+    ) -> EvalResult<'tcx> {
+        trace!("validate scalar: {:#?}, {:#?}, {:#?}", value, size, scalar);
+        use std::ops::RangeBounds;
+        use std::fmt::Debug;
+        fn in_range<'tcx>(value: Scalar, size: Size, bound: impl RangeBounds<u128> + Debug) -> EvalResult<'tcx, bool> {
+            match value {
+                Scalar::Bits {..} => {
+                    let bits = value.to_bits(size)?;
+                    Ok(bound.contains(&bits))
+                },
+                Scalar::Ptr(_) => {
+                    // we cannot reason about Ptr numeric ranges other than zero
+                    if bound.start_bound() == Bound::Included(&0) && bound.end_bound() == Bound::Included(&0) {
```

Comment:
So this says that the range has to be just (0,0) for the pointer to invalidate the range? When would such ranges even occur? I have no idea what `scalar.valid_range` is for e.g. pointer types.

Reply:
yes, I just did the most conservative thing i could think of. Usually you have `0..` or `1..` as the valid range.

Reply:
But this means usually, this will never fire?

Also, raw pointers can be null. Does this check to make sure we're talking about a reference?

And finally, what about alignment?

Reply:
> But this means usually, this will never fire?

Yea, unless you define a clike enum with 2^32-1 variants without it having one where the discriminant is `0`

> Also, raw pointers can be null. Does this check to make sure we're talking about a reference?

Oh that's not the issue, if you have a `0` you do not get into the `Scalar::Ptr` arm, this is an abstract pointer. You'd be in the `Scalar::Bits` arm.

Also we're not checking types here, we're checking layout. And the layout of a raw pointer allows having a `0`.

> And finally, what about alignment?

That would require us to look at the type. I can add more checks, I just wanted to start with the very basic layout checks.

Reply:
Oh, that's the `Scalar` miri value thing, not layout.

> Yea, unless you define a clike enum with 2^32-1 variants without it having one where the discriminant is 0

Now I am confused. The only way this fires, from what I can see, is having an enum where `0` is the *only* variant, and then having a pointer value at that enum type.

> That would require us to look at the type.

I see, makes sense.

Reply:
> Now I am confused. The only way this fires, from what I can see, is having an enum where 0 is the only variant, and then having a pointer value at that enum type.

you're right of course. I swapped the logic in my head.

Reply:
Okay, can we have a test please? :)

Reply:
`valid_range` is a wraparound range, which complicates this logic. I think what we can do is split it into two inclusive ranges, when `start >= end`, i.e. `0..=start` and `end..=(!0 >> (128 - size.bits()))`.

Reply:
Your logic differs in that you have a `start > end` condition, which is how you get the `0..=0` range, but the validity ranges are exclusive, so.... - hang on a second, I just checked the source and it has `valid_range: RangeInclusive<u128>,`

I'm getting more confused by the minute, if they're inclusive ranges, how are we getting `0..=0` in the first place? Or is the `valid_range` `1..=0`, (which is the full range of the type, just rotated)? I think it's what you get in something like `Option<&T>`.


Either way, it'd be simpler if you always created two complete `RangeInclusive`'s rather than things like `..=hi` and `lo..`.

Reply:
`valid_range` is indeed `1..=0` for some full-range types.

You get `0..=0` for `enum Foo { Bar = 0 }`

> Either way, it'd be simpler if you always created two complete RangeInclusive's rather than things like ..=hi and lo...

indeed, no generics means I can use a closure

Reply:
`0..=0` meaning exactly one value, `0`, makes sense, but at the same time, it means your check for pointers *doesn't* make sense, since they'd never have the `0..=0` range, would they?

Reply:
The pointer check is checking the opposite. It checks if you have a type `enum Foo { Bar = 0 }` but the value is a pointer. So if you transmuted a `&1` to a `Foo`, then this check triggers. Transmuting a `&1` to an `enum Foo { Bar = 42 };` is uncheckable at compile-time, because the pointer might actually point to address `42`

Reply:
That's confusing. Why not check that the range includes all the possible values of the pointer? Sadly, this doesn't work very well with splitting into two ranges, I don't think, so the splitting should probably only be done for `Scalar::Bits`.

A `Scalar::Pointer` could have *any* address, so assuming it's any *specific*  one, is generally a bad idea.